packages/grid/x-data-grid/src/utils/createSelector.ts

```diff
@@ -48,12 +48,99 @@ const missingInstanceIdWarning = buildWarning([
+  d?: Function,
+  e?: Function,
+  f?: Function,
+  ...rest: any[]
+) => {
+  if (rest.length > 0) {
+    throw new Error('Unsupported number of selectors');
+  }
+
+  let selector: any;
+
+  if (a && b && c && d && e && f) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      const vd = d(state, instanceId);
+      const ve = e(state, instanceId);
+      return f(va, vb, vc, vd, ve);
+    };
+  } else if (a && b && c && d && e) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      const vd = d(state, instanceId);
+      return e(va, vb, vc, vd);
+    };
+  } else if (a && b && c && d) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      return d(va, vb, vc);
+    };
+  } else if (a && b && c) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      return c(va, vb);
+    };
+  } else if (a && b) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      return b(va);
+    };
+  } else {
+    throw new Error('Missing arguments');
+  }
+
+  // We use this property to detect if the selector was created with createSelector
+  // or it's only a simple function the receives the state and returns part of it.
+  selector.acceptsApiRef = true;
+
+  return selector;
+}) as unknown as CreateSelectorFunction;
```

Comment:
Will doing it in a more generic way make it less performant?
(sorry if it's already discussed, I couldn't find a conversation)
Something like:
```suggestion
export const createSelector = ((
  ...fns: Function[]
) => {
  if (fns.length > 6) {
    throw new Error('Unsupported number of selectors');
  }

  let selector: any;

  if (fns.length >= 2) {
    selector = (stateOrApiRef: any, instanceIdParam: any) => {
      const isAPI = isAPIRef(stateOrApiRef);
      const instanceId =
        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
      return fns[fns.length - 1](
        ...fns.slice(0, -1).map((fn: Function) => fn(state, instanceId)),
      );
    };
  } else {
    throw new Error('Missing arguments');
  }

  // We use this property to detect if the selector was created with createSelector
  // or it's only a simple function the receives the state and returns part of it.
  selector.acceptsApiRef = true;

  return selector;
}) as unknown as CreateSelectorFunction;
```

Reply:
Yeah, it's about performance. The selectors are used everywhere, so e.g. in filtering if we have 100,000 rows with each iteration calling 10 selectors, it makes 1,000,000 selector calls right there. At those numbers, small differences matter.

Reply:
Does it actually bring some performance improvement though? 🤔 

Just out of curiousity, I did some benchmarks and here's a comparison on different data size and searching `am` text string on `Trader Name` column and clearing it, multiple times.

### v8 comparison, rows: 100,000

| Generic createSelector | Current version |
| --- | --- |
| <img width="322" alt="image" src="https://github.com/mui/mui-x/assets/12609561/0b1d7c6d-8d70-4c8f-ad45-245ce2abfb9d"> | <img width="338" alt="image" src="https://github.com/mui/mui-x/assets/12609561/89cb5c81-6266-4ed1-8192-e20fd9f3d50b"> |

### SpiderMonkey comparison, rows: 100,000

| Generic createSelector | Current version |
| --- | --- |
| <img width="322" alt="image" src="https://github.com/mui/mui-x/assets/12609561/f79b5634-061c-43ad-9af6-25821814c2f8"> | <img width="338" alt="image" src="https://github.com/mui/mui-x/assets/12609561/345a730b-460d-4bb2-bd8c-b395c0b882ac"> |


### v8 comparison, rows: 1,000,000

| Generic createSelector | Current version |
| --- | --- |
| <img width="323" alt="image" src="https://github.com/mui/mui-x/assets/12609561/aeed6ccb-c799-494d-a330-84766fc0b0ba"> | <img width="340" alt="image" src="https://github.com/mui/mui-x/assets/12609561/c6c5e62d-5820-4140-91c7-002e00c8747f"> |

### SpiderMonkey comparison, rows: 1,000,000

| Generic createSelector | Current version |
| --- | --- |
| <img width="322" alt="image" src="https://github.com/mui/mui-x/assets/12609561/3695c087-aa23-454a-a903-8cb45dc8ee2f"> | <img width="338" alt="image" src="https://github.com/mui/mui-x/assets/12609561/6c53bc8e-1a2c-473b-94fa-9afc901836bf"> |

The thing I'm thinking about:
1. Is the performance boost worth it to compromise the readability a bit, IMO, the generic variant looks cleaner and it's easier to support more number of selectors
2. Uptil 100,000 rows, the difference is negligible, some times the generic version performed better. Do you think this much number of rows (1,000,000) is a realistic use case on the client side? To me, it'd make more sense to handle such use-cases on the server side, or at least use pagination on the client side.

Reply:
Here are my thoughts:
1. The code is more verbose but not necessarily less readable, the amount of complexity introduced is low.
2. I think it's a rare use-case, but for lower end devices the number of rows where it will pay off will be lower, so I would keep it. Note, you mention pagination but pagination doesn't affect filtering.
3. It's also easier to debug code that uses selectors with the performant version than with the generic version. When you're stepping in that code, you can clearly see all the selectors & their arguments laid out.
4. I don't have a strong opinion on this, if you feel we should take the generic version I'll switch to it.

Reply:
No strong opinion on this one, but IMO the verbose version is more redable